cmake_pc_hooks/_cmake.py

```diff
@@ -0,0 +1,305 @@
+
+
+def get_cmake_command(cmake_names=None):  # pragma: nocover
+    """
+    Get the path to a CMake executable on the PATH or in the virtual environment.
+
+    Args:
+        cmake_names (:obj:`list` of :obj:`str`): Names for the CMake command
+            Defaults to ['cmake', 'cmake3']
+    """
+    if not cmake_names:
+        cmake_names = ['cmake', 'cmake3']
+
+    for cmake in cmake_names:
+        with Path(os.devnull).open(mode='w', encoding='utf-8') as devnull:
+            try:
+                sp.check_call([cmake, '--version'], stdout=devnull, stderr=devnull)
+                return [shutil.which(cmake)]
+            except (OSError, sp.CalledProcessError):
+                pass
+
+            # CMake not in PATH, should have installed Python CMake module
+            # -> try to find out where it is
+            python_executable = Path(sys.executable)
+            try:
+                root_path = Path(os.environ['VIRTUAL_ENV'])
+                python = python_executable.name
+            except KeyError:
+                root_path = python_executable.parent
+                python = python_executable.name
+
+            search_paths = [root_path, root_path / 'bin', root_path / 'Scripts']
+
+            # First try executing CMake directly
+            for base_path in search_paths:
+                try:
+                    cmake_cmd = base_path / cmake
+                    sp.check_call([cmake_cmd, '--version'], stdout=devnull, stderr=devnull)
+                    return [cmake_cmd]
+                except (OSError, sp.CalledProcessError):
+                    pass
+
+            # That did not work: try calling it through Python
+            for base_path in search_paths:
+                try:
+                    cmake_cmd = [python, base_path / 'cmake']
+                    sp.check_call([*cmake_cmd, '--version'], stdout=devnull, stderr=devnull)
+                    return cmake_cmd
+                except (OSError, sp.CalledProcessError):
+                    pass
+
+    # Nothing worked -> give up!
+    return None
+
+
+class CMakeCommand:
+    """Class used to encapsulate all CMake related functionality."""
+
+    DEFAULT_BUILD_DIR = '.cmake_build'
+
+    def __init__(self, cmake_names=None):
+        """
+        Initialize a CMakeCommand object.
+
+        Args:
+            cmake_names (list[str] | None): List of possible names for the CMake executable
+        """
+        self.command = get_cmake_command(cmake_names)
+        self.source_dir = None
+        self.build_dir = None
+        self.cmake_args = ['-DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL=ON']
+
+    def add_cmake_arguments_to_parser(self, parser):
+        """Add CMake options to an argparse.ArgumentParser."""
+        parser.add_argument('-S', '--source-dir', type=str, help='Path to build directory', default='.')
+        parser.add_argument('-B', '--build-dir', action='append', type=str, help='Path to build directory')
```

Comment:
Shouldn't be `action=append`

Reply:
Not for the source directory. I am only allowing a single source directory. I allow for multiple build directories so that you can have a configuration that works on multiple computers where build directories are not all named the same.

This mainly stemmed from my usage where I might have multiple build directories and I want the hooks to use the build directories that exists and not have to worry about changing the pre-commit configuration when I delete one of those build directories.

Reply:
Oh yeah, I forgot about that workflow. But then we have to be careful with presets since that is assuming that you want a specific configuration, not any configuration.

And just to be on the safeside, when you run a configure step again with no cache variables passed, it does not run a re-configure right? E.g. when you run it with trace and file-api.

Reply:
If nothing changed I think it won't. But otherwise it will use the existing cache and the preset is probably stored in the cache to begin with so that should be fine.

Reply:
In the case of using --preset, you can always specify only a single build directory (or none and it will create one*)

*once I add the option to ignore the automatic search ;-)